File: java-tomcat/src/main/java/com/ederfmatos/javatomcat/usecase/GetTransaction.java
package com.ederfmatos.javatomcat.usecase;

import com.ederfmatos.javatomcat.entity.Transaction;
import com.ederfmatos.javatomcat.enums.Currency;
import com.ederfmatos.javatomcat.exceptions.TransactionNotFoundException;
import com.ederfmatos.javatomcat.repository.TransactionRepository;
import java.math.BigDecimal;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class GetTransaction {
    private final TransactionRepository transactionRepository;

    public Output execute(Input input) {
        return transactionRepository.findById(input.id)
                .map(Output::create)
                .orElseThrow(TransactionNotFoundException::new);
    }

    public record Input(String id) {
    }

    public record Output(String id, String description, BigDecimal amount, BigDecimal totalAmount, Currency currency) {
        public static Output create(Transaction transaction) {
            return new Output(
                    transaction.id(),
                    transaction.description(),
                    transaction.amount(),
                    transaction.totalAmount(),
                    transaction.currency()
            );
        }
    }
}

File: kotlin-coroutines/src/main/kotlin/com/ederfmatos/kotlincoroutines/usecase/GetTransaction.kt
package com.ederfmatos.kotlincoroutines.usecase

import com.ederfmatos.kotlincoroutines.enums.Currency
import com.ederfmatos.kotlincoroutines.exception.TransactionNotFoundException
import com.ederfmatos.kotlincoroutines.repository.TransactionRepository
import org.springframework.stereotype.Component
import java.math.BigDecimal

@Component
class GetTransaction(private val transactionRepository: TransactionRepository) {

    suspend fun execute(input: Input): Output {
        val transaction = transactionRepository.findById(input.id) ?: throw TransactionNotFoundException()
        return Output(
            id = transaction.id,
            description = transaction.description,
            amount = transaction.amount,
            currency = transaction.currency,
            totalAmount = transaction.totalAmount
        )
    }

    data class Input(val id: String)
    data class Output(
        val id: String,
        val description: String,
        val amount: BigDecimal,
        val totalAmount: BigDecimal,
        val currency: Currency,
    )
}